README.md
# DecisiveML
[![Build Status](https://travis-ci.com/decisivealpha/decisiveml.svg?branch=master)](https://travis-ci.com/decisivealpha/decisiveml)
[![PyPi](https://img.shields.io/pypi/v/decisiveml)](https://pypi.org/project/decisiveml/)
[![Python](https://img.shields.io/pypi/pyversions/decisiveml.svg)](https://pypi.org/project/decisiveml/)
[![Discord Chat](https://img.shields.io/discord/594866572104040498.svg)](https://discord.gg/zQRSUhF)

DecisiveML enables end-to-end research and trade execution with machine learning

## Install

`pip install decisiveml`

## Develop

Dependency management uses [poetry](https://github.com/python-poetry/poetry), test suite is [nose](https://github.com/nose-devs/nose), and code style is [black](https://github.com/psf/black).

To develop:

1. Initialize with `make install`
2. Run tests with: `make test`

## Examples

See [notebooks](https://github.com/decisivealpha/DecisiveML/tree/master/notebooks) for more

### Trend Scanning

See [docs](https://github.com/decisivealpha/DecisiveML/tree/master/docs/trend_scanning.md) for more on Trend Scanning

```python
import decisiveml as dml
import numpy as np
import pandas as pd

# Generate random price history
raw_frame = pd.DataFrame(index=pd.date_range("2018-12-01", "2019-07-01"))
raw_frame["equity"] = (np.random.randn(raw_frame.shape[0]) / 100).cumsum()
raw_frame["close"] = 48.76 * (1 + raw_frame["equity"])

# Use a cross-over signal as entry
df = raw_frame["2019-01-01":"2019-06-01"].copy()
df["mavg"] = df.close.rolling(10).mean()
m_crossabove = (df.close.shift(1) < df.mavg) & (df.close > df.mavg)
m_crossbelow = (df.close.shift(1) > df.mavg) & (df.close < df.mavg)
df["entry"] = df[m_crossabove | m_crossbelow].close

# Calculate trendscanning
trend = dml.getBinsFromTrend(
    molecule=df["entry"].dropna().index, close=df.close, span=[22, 44, 11],
)
```

### Monte Carlo

See [docs](https://github.com/decisivealpha/DecisiveML/tree/master/docs/monte_carlo.md) for more on Monte Carlo

```python
import decisiveml as dml

# set up margin requirements for instrument
margin = 5000
start_date = datetime.date(2016, 1, 1)
end_date = datetime.date(2018, 1, 1)

# list of trades
trades_list = random.sample(range(-2000, 2300), 100)

# Initialize
mc = dml.MonteCarlo(trades_list)

# We will sample with replacement the number of trades per year
# so we need the start and end date to determine how many trades at in a year on average
mc.settings(margin, start_date, end_date)

# Test different levels of equity starting at this value
trial_starting_equity = int(margin * 1.5)

# Run the Monte Carlo
results = mc.run(trial_starting_equity)

# Results:
# ========
# Recommend a starting equity of 15000.0, which has 5.9% Risk-of-Ruin, 81% Probability-of-Profit and a 1.65 Returns/Drawdown Ratio
# Risk Assessment: FAILED
# MC-Drawdown: 30.2% MC-1.5x-DD: 45.2%
# Average monthly net profit: 307.8
```

## Support

- Chat in [Discord](https://discord.gg/zQRSUhF)
- Submit [issues](https://github.com/decisivealpha/DecisiveML/issues)
- Submit [pull requests](https://github.com/decisivealpha/DecisiveML/pulls)

Released under [BSD-3-Clause](https://github.com/decisivealpha/DecisiveML/blob/master/LICENSE.txt) license


CHANGELOG.md
# CHANGELOG

## [0.0.3] - 2020-05-08
### Added
- Basic indicators
- Helper methods for trading date offsets

## [0.0.2] - 2020-05-03
### Added
- Documentation for Monte Carlo and Trend Scanning
- Recommendation method for Monte Carlo

## [0.0.1] - 2020-04-15
### Added
- Module to calculate Monte Carlo and Risk of Ruin
- Trend Scanning module and notebook


tests/test_ml.py
from nose.tools import eq_
import numpy as np
import pandas as pd
import decisiveml as dml


def test_trendscanning():
    """Test trend scanning"""
    # Generate random price history
    raw_frame = pd.DataFrame(index=pd.date_range("2018-12-01", "2019-07-01"))
    np.random.seed(0)
    raw_frame["equity"] = (np.random.randn(raw_frame.shape[0]) / 100).cumsum()
    raw_frame["close"] = 48.76 * (1 + raw_frame["equity"])

    # Use a cross-over signal as entry
    df = raw_frame["2019-01-01":"2019-06-01"].copy()
    df["mavg"] = df.close.rolling(10).mean()
    m_crossabove = (df.close.shift(1) < df.mavg) & (df.close > df.mavg)
    m_crossbelow = (df.close.shift(1) > df.mavg) & (df.close < df.mavg)
    df["entry"] = df[m_crossabove | m_crossbelow].close

    # Calculate trendscanning
    trend = dml.getBinsFromTrend(
        molecule=df["entry"].dropna().index, close=df.close, span=[22, 44, 11],
    )
    eq_(trend.iloc[1].bin, 1)
    eq_(trend.iloc[-1].bin, -1)
    eq_(trend.bin.sum(), 12)


tests/test_helpers.py
#!/usr/bin/env python3
import unittest
from nose.tools import eq_
import pandas as pd
import os
import decisiveml as dml

TEST_DIR = os.path.dirname(os.path.realpath(__file__))


class TestTradingHolidays(unittest.TestCase):
    def setUp(self):
        self.short_holiday = pd.read_csv(
            os.path.join(
                TEST_DIR, "test_data/holidays/Holidays_Daily_2019-01-02_2020-02-20.csv"
            ),
            index_col="Date",
        )

    def test_trading_holidays_in_range(self):
        """Test that we can remove our own holidays"""
        holidays = dml.trading_holidays_in_range(start="2018-12-01", end="2018-12-31")
        eq_(len(holidays), 2)
        eq_(holidays[0], "2018-12-25")
        eq_(holidays[1], "2019-01-01")
        eq_("2018-12-05" not in holidays, True)

    def test_trading_holidays_in_range_with_short_holiday(self):
        """Test using HOLIDAY data and fix missing dates in pandas_market_calendar"""
        df = self.short_holiday
        holidays = dml.trading_holidays_in_range(start=df.index[0], end=df.index[-1])
        eq_("2020-01-20" in holidays, True)
        eq_("2020-02-17" in holidays, True)

    def test_tradingday_offset_short(self):
        """Test that we can offset a business day through a holiday"""
        offsets = dml.tradingday_offset(
            start="2017-04-10",
            end="2017-04-20",
            dates=["2017-04-17"],
            offsets=-1,
            roll="backward",
        )
        eq_(offsets[0], pd.to_datetime("2017-04-13"))

    def test_tradingday_offset_long(self):
        """Test that we can offset a business day through a holiday"""
        offsets = dml.tradingday_offset(
            start="2017-01-01",
            end="2017-05-01",
            dates=["2017-01-01", "2017-04-17"],
            offsets=-1,
            roll="backward",
        )
        eq_(offsets[0], pd.to_datetime("2016-12-29"))
        eq_(offsets[1], pd.to_datetime("2017-04-13"))


decisiveml/indicators.py
#!/usr/bin/env python3
import numpy as np
import pandas as pd
import logging
from decisiveml.helpers import tradingday_offset

logger = logging.getLogger(__name__)


def indicator_volatility_daily(df_daily, price_col="close"):
    """Create rolling volatility using EWM of 36 which matches a rolling stdev of 25

    Args:
        df_daily (frame): contains "close" column
        price_col (str): "close" is default

    Returns:
        frame: index = index of df_daily, adds colum called vol36

    """
    df = df_daily[[price_col]].copy()
    df["pct_returns"] = df[price_col].pct_change()
    df["vol36"] = df.pct_returns.ewm(span=36, adjust=False).std()
    df.drop(["pct_returns", price_col], axis=1, inplace=True)
    df.dropna(inplace=True)
    return df


def indicator_bollinger(df_t, lookback=20):
    """Creates columns for bollinger channel.  Requires columns "high" and "low" in intraday_df

    Args:
        df (pd.DataFrame): OHLCV intraday dataframe, only needs close
        lookback (:obj:int, optional): rolling window.  Default is 20

    Returns:
        pd.DataFrame: columns are:
            - bollinger_high
            - bollinger_low
    """
    df = df_t[["close"]].copy()
    df["bb_ma"] = df.close.rolling(window=lookback).mean()
    df["bb_std"] = df.close.rolling(window=lookback).std()
    df["bollinger_high"] = df.bb_ma + df.bb_std
    df["bollinger_low"] = df.bb_ma - df.bb_std
    df = df.drop(["close", "bb_ma", "bb_std"], axis=1)
    return df


def indicator_volbands(df_t, lookback=20, multiplier=0.3):
    """Create bands around a mean based on volatility

    Args:
        df_t (frame): requires "close" and "vol36" -- which is daily volatility
        lookback (`obj`:int, optional): lookback for moving average
        multiplier (`obj`:float, optional): multiplier for the daily volatility. 0.3 works well for about a 30T timeframe

    Returns:
        frame: vol_high and vol_low, which is the daily volatility above a moving average
    """
    df = df_t[["close", "vol36"]].copy()
    df["vol_ma"] = df.close.rolling(window=lookback).mean()
    df["vol_width"] = df.close * df.vol36 * multiplier
    df["vol_high"] = df.vol_ma + df.vol_width
    df["vol_low"] = df.vol_ma - df.vol_width
    return df[["vol_high", "vol_low"]]


def indicator_donchian(df, lookback=20):
    """Creates columns for donchian channel.  Requires columns "high" and "low" in intraday_df

    Args:
        df (pd.DataFrame): OHLCV intraday dataframe, only needs two columns:
            - high: used for hold-high
            - low: used for hold-low
        lookback (:obj:int, optional): rolling window.  Default is 20

    Returns:
        pd.DataFrame: columns are:
            - donchian_high
            - donchian_low

    Example:
        >>> df = processed_df.copy()
        >>> donchian = dvind.indicator_donchian(df, lookback=1380*3)
        >>> df = pd.merge(df, donchian, left_index=True, right_index=True)
    """
    df = df[["high", "low"]].copy()
    df = df.dropna()
    df["donchian_high"] = df.high.rolling(lookback).max()
    df["donchian_low"] = df.high.rolling(lookback).min()
    df = df.drop(["high", "low"], axis=1)
    return df


def indicator_atr(df, lookback=20):
    """Creates average true range

    Args:
        df (pd.DataFrame): OHLCV intraday dataframe, only needs high, low, close
        lookback (:obj:int, optional): rolling window.  Default is 20

    Returns:
        pd.DataFrame: columns with true_range and atr

    Example:
    """
    df = df[["high", "low", "close"]].copy()

    # true range
    df["highlow"] = abs(df["high"] - df["low"])
    df["highclose"] = abs(df["high"] - df["close"].shift(1))
    df["lowclose"] = abs(df["low"] - df["close"].shift(1))
    df["true_range"] = df[["highlow", "highclose", "lowclose"]].max(axis=1)

    # average
    df["atr"] = df["true_range"].rolling(lookback).mean()

    return df[["atr", "true_range"]]


def indicator_pp_daily(intraday_df):
    """Create a daily dataframe from EST data

    Args:
        intraday_df (pd.DataFrame): EST, left-indexed, i.e. starts at YYYY-MM-DD 18:00

    Returns:
        pd.DataFrame: index are pd.Timestamp of session-closing dates for the session pivot point,
            i.e. 2020-02-03 refers to 2020-02-02 18:00 to 2020-02-03 17:00

    Example:
        >>> ppframe_daily = indicator_pp_daily(df_all)
        >>> assert trading_gaps(ppframe_daily.index).empty
        >>> ppframe = set_index_to_intraday_start(ppframe_daily)
        >>> df_all[ppframe.columns] = ppframe
        >>> df_all[ppframe.columns] = df_all[ppframe.columns].fillna(method="ffill")
    """

    # Convert EST into session closing dates (no HH:MM)
    df = intraday_df.resample("24H", label="right", base=18).agg(
        {"open": "first", "high": "max", "low": "min", "close": "last", "volume": "sum"}
    )
    df.dropna(inplace=True)
    df.index = df.index.date

    # Add one day to the end before we shift
    start = intraday_df.index[0].strftime("%Y-%m-%d")
    end = intraday_df.index[-1].strftime("%Y-%m-%d")
    additional_day = tradingday_offset(
        start=start, end=end, dates=df.index[-1], offsets=1, roll="forward"
    )
    df.loc[additional_day] = np.nan

    # Convert to timestamp and shift
    df.index = pd.to_datetime(df.index)
    prev_day = df.shift(1)

    # Calculate pivot points use the previous day values
    pivot_df = pd.DataFrame()
    pivot_df["pp"] = (prev_day["high"] + prev_day["low"] + prev_day["close"]) / 3
    pivot_df["r1"] = 2 * pivot_df["pp"] - prev_day["low"]
    pivot_df["s1"] = 2 * pivot_df["pp"] - prev_day["high"]
    pivot_df["r2"] = pivot_df["pp"] + pivot_df["r1"] - pivot_df["s1"]
    pivot_df["s2"] = pivot_df["pp"] + pivot_df["s1"] - pivot_df["r1"]
    pivot_df["r3"] = pivot_df["pp"] - pivot_df["s2"] + pivot_df["r2"]
    pivot_df["s3"] = pivot_df["pp"] - pivot_df["r2"] + pivot_df["s2"]

    return pivot_df


decisiveml/montecarlo.py
import pandas as pd
import random
import statistics
import logging

logger = logging.getLogger(__name__)

# MonkeyPatch Python 3.6 choices into random 3.5.5
import bisect as _bisect
import itertools as _itertools


def choices(population, weights=None, cum_weights=None, k=1):
    """Return a k sized list of population elements chosen with
replacement.
    If the relative weights or cumulative weights are not specified,
    the selections are made with equal probability.
    """
    if cum_weights is None:
        if weights is None:
            total = len(population)
            return [population[int(random.random() * total)] for i in range(k)]
        cum_weights = list(_itertools.accumulate(weights))
    elif weights is not None:
        raise TypeError("Cannot specify both weights and cumulative weights")
    if len(cum_weights) != len(population):
        raise ValueError("The number of weights does not match the population")
    bisect = _bisect.bisect
    total = cum_weights[-1]
    hi = len(cum_weights) - 1
    return [population[bisect(cum_weights, random() * total, 0, hi)] for i in range(k)]


random.choices = choices


class ExcessiveBaseEquity(Exception):
    pass


class MonteCarlo(object):
    def __init__(self, trades_list):
        self.trades_list = trades_list
        self.num_trades_total = len(self.trades_list)
        self.num_trades_per_year = None
        self.ruin_equity = None
        self.runs = None

        self._MONTECARLO_RUNS = 2500
        logger.info(
            "Initialize \t| Trades: {} \t| MC Runs: {}".format(
                self.num_trades_total, self._MONTECARLO_RUNS
            )
        )

    def settings(self, ruin_equity, start_date, end_date):
        self._set_ruin_equity(ruin_equity)
        self._set_trades_per_year(start_date, end_date)
        logger.info(
            "Settings \t| Ruin: {} \t| Trades Per Year: {}".format(
                self.ruin_equity, self.num_trades_per_year
            )
        )

    def _set_ruin_equity(self, ruin_equity):
        self.ruin_equity = ruin_equity

    def _set_trades_per_year(self, start_date, end_date):
        td = end_date - start_date
        self.num_trades_per_year = int(self.num_trades_total * 365 / td.days)
        logger.debug("TimeDelta: {} {} trades/yr".format(td, self.num_trades_per_year))

    def _random_trade(self, starting_equity):
        assert self.num_trades_per_year
        assert self.ruin_equity

        trades = random.choices(self.trades_list, k=self.num_trades_per_year)
        # logger.debug("{} {}".format(len(trades), trades))

        # Check for ruin at any point in the trades list
        is_ruined = 0
        equity = starting_equity
        for trade in trades:
            equity = equity + trade
            if equity < self.ruin_equity:
                is_ruined = 1
                break

        stats = {
            "profit": sum(trades),
            "returns_pct": int(
                100 * ((starting_equity + sum(trades)) / starting_equity - 1)
            ),
            "drawdown_pct": self._drawdown(starting_equity, trades),
            "is_ruined": is_ruined,
            "is_profitable": 1 if sum(trades) >= 0 else 0,
        }
        try:
            stats["returns_per_drawdown"] = stats["returns_pct"] / stats["drawdown_pct"]
        except ZeroDivisionError:
            stats["returns_per_drawdown"] = 0
        logger.debug(stats)
        return stats

    def _drawdown(self, starting_equity, trades):
        """Returns the maximum drawdown in a set of trades"""
        equity = starting_equity
        hwm = starting_equity
        max_drawdown_pct = 0
        drawdown_pct = 0

        for trade in trades:
            equity = equity + trade

            # Set High Water Mark
            if equity > hwm:
                hwm = equity

            # Look for drawdown
            if equity < hwm:
                drawdown_pct = 100 * (1 - (equity / hwm))
                if drawdown_pct > max_drawdown_pct:
                    max_drawdown_pct = drawdown_pct

            # logger.debug("{} eq:{} \t hwm:{} \t dd:{} mdd:{}".format(trade, equity, hwm, drawdown_pct, max_drawdown_pct))
        return max_drawdown_pct

    def _median_stats_run(self, starting_equity):
        montecarlo = {}
        median_montecarlo = {}

        # runs
        for _ in range(self._MONTECARLO_RUNS):

            # make the list for every key
            stats = self._random_trade(starting_equity)
            for k, v in stats.items():
                if k not in montecarlo.keys():
                    montecarlo[k] = []
                montecarlo[k].append(v)

        # run statistics on all the lists of every key
        for k, v in montecarlo.items():
            # ignore non-median stats
            if k != "is_ruined" or k != "is_profitable":
                median_montecarlo[k] = statistics.median(montecarlo[k])

        logger.debug((montecarlo["profit"]))
        logger.debug((montecarlo["is_ruined"]))
        logger.debug(sum(montecarlo["is_ruined"]))
        median_montecarlo["is_ruined"] = (
            100 * sum(montecarlo["is_ruined"]) / self._MONTECARLO_RUNS
        )
        median_montecarlo["is_profitable"] = (
            100 * sum(montecarlo["is_profitable"]) / self._MONTECARLO_RUNS
        )
        median_montecarlo["equity"] = starting_equity

        # calculate risk of ruin
        logger.debug("Median {}: {}".format(starting_equity, median_montecarlo))

        return median_montecarlo

    def run(self, base_equity, steps=11):
        """Create the results for the MonteCarlo, adding equity to the
        base_equity

        Args:
            base_equity (int): starting equity to add to
            steps (:obj:`int`, optional). Default is 11 runs.

        Returns:
            pd.DataFrame: results for each run with various starting equities

        Example:
            >>> mc = dvm.MonteCarlo(trade_list)
            >>> start_date = trade_list.index[0].to_pydatetime()
            >>> end_date = trade_list.index[-1].to_pydatetime()
            >>> mc.settings(ruin_equity=5000, start_date=start_date, end_date=end_date)
            >>> results = mc.run(base_equity=starting_equity)

        """
        step_size = int(base_equity / 4)
        end_eq = base_equity + step_size * steps
        starting_equities_list = range(base_equity, end_eq, step_size)
        results = self._run_equity_list(starting_equities_list)
        df = pd.DataFrame(results)
        return df

    def _run_equity_list(self, starting_equities_list):
        runs = []
        for starting_equity in starting_equities_list:
            runs.append(self._median_stats_run(starting_equity))
        self.runs = runs
        return runs

    def best_run(self, target_risk_of_ruin_pct=10):
        assert self.runs
        for run in self.runs:
            if run["is_ruined"] < target_risk_of_ruin_pct:
                return run
        return None

    def recommendation(self, start_date, end_date):
        """Create a recommendation from the best run

        Args:
            start_date (datetime): trade start date
            end_date (datetime): trade end date

        Returns:
            pd.Series: recommendation categories

        Example:
            >>> mc = dvm.MonteCarlo(trade_list)
            >>> start_date = trade_list.index[0].to_pydatetime()
            >>> end_date = trade_list.index[-1].to_pydatetime()
            >>> mc.settings(ruin_equity=5000, start_date=start_date, end_date=end_date)
            >>> results = mc.run(base_equity=starting_equity)
            >>> my_rec = mc.recommendation(start_date, end_date)
        """

        # Get the recommended values
        best = self.best_run()
        my_rec = pd.Series(best)

        if my_rec.empty:
            raise ExcessiveBaseEquity(f"No best run found")

        # Determine result
        if my_rec["is_ruined"] > 10 or my_rec["returns_per_drawdown"] < 2.0:
            logger.info("MonteCarlo Risk Assessment: FAILED")
            my_rec["is_pass"] = False
        else:
            logger.info("MonteCarlo Risk Assessment: PASSED")
            my_rec["is_pass"] = True

        # Add additional calculations
        my_rec["start_date"] = start_date
        my_rec["end_date"] = end_date
        my_rec["months"] = (end_date - start_date).days / 30
        my_rec["avg_monthly_profit"] = my_rec["profit"] / my_rec["months"]

        return my_rec


decisiveml/trendscanning.py
import pandas as pd
import numpy as np
import statsmodels.api as sm1


def tValLinR(close):
    """tValue from a linear trend via SNIPPET 5.1 T-VALUE OF A LINEAR TREND

    Args:
        close (Series): close prices to search through

    Returns:
        float: t-value

    Example:
        >>> lookforward = 5
        >>> start = 0
        >>> stop = stop + lookforward
        >>> df1 = close.iloc[start:stop]
        >>> tValLinR(df1.values)

    """
    x = np.ones((close.shape[0], 2))
    x[:, 1] = np.arange(close.shape[0])
    ols = sm1.OLS(close, x).fit()
    return ols.tvalues[1]


def getBinsFromTrend(molecule, close, span):
    """Derive labels from the sign of t-value of linear trend via SNIPPET 5.2
    IMPLEMENTATION OF THE TREND-SCANNING METHOD

    Args:
        molecule (DatetimeIndex): start times of the event
        close (Series): close prices of your large df
        span (args for range): list: [start, stop, step]

    Returns:
        pd.DataFrame: trends
            - t1: End time for the identified trend
            - tVal: t-value associated with the estimated trend coefficient
            - bin: Sign of the trend

    Example:
        >>> df["mavg"] = df.close.rolling(10).mean()
        >>> df["entry"] = df[(df.close.shift(1) < df.mavg) & (df.close > df.mavg)].close
        >>> getBinsFromTrend(molecule=df["entry"].dropna().index, close=df.close, span=[5, 20, 5])
    """

    out = pd.DataFrame(index=molecule, columns=["t1", "tVal", "bin"])
    hrzns = range(*span)

    for dt0 in molecule:
        df0 = pd.Series()
        iloc0 = close.index.get_loc(dt0)
        if iloc0 + max(hrzns) > close.shape[0]:
            continue

        for hrzn in hrzns:
            dt1 = close.index[iloc0 + hrzn - 1]
            df1 = close.loc[dt0:dt1]
            df0.loc[dt1] = tValLinR(df1.values)

        dt1 = df0.replace([-np.inf, np.inf, np.nan], 0).abs().idxmax()
        out.loc[dt0, ["t1", "tVal", "bin"]] = (
            df0.index[-1],
            df0[dt1],
            np.sign(df0[dt1]),
        )  # prevent leakage

    out["t1"] = pd.to_datetime(out["t1"])
    out["bin"] = pd.to_numeric(out["bin"], downcast="signed")

    return out.dropna(subset=["bin"])


decisiveml/helpers.py
#!/usr/bin/env python3
import pandas as pd
import numpy as np
import pandas_market_calendars as mcal


def set_index_to_intraday_start(daily_df):

    """Convert daily dataframe into intraday start times.

    For example, the "2020-02-02" session will be indexed to "2020-02-01 18:00"

    Args:
        daily_df (frame): daily indicator with timestamp index in YYYY-MM-DD format in EST

    Returns:
        frame: index set to intraday start time in EST. You will need to ffill
            the result if merging into an intraday dataframe

    Example:
        >>> ppframe_daily = indicator_pp_daily(df_ins)["2020-02-02":]
        >>> ppframe = set_index_to_intraday_start(ppframe_daily)
        >>> df_ins[ppframe.columns] = ppframe
        >>> df_ins[ppframe.columns] = df_ins[ppframe.columns].fillna(method="ffill")

        If you want to go the other way, you can do:
        >>> my_dates = list(df.index.date)
        >>> session_end = tradingday_offset(
        >>>     start=my_dates[0],
        >>>     end=my_dates[-1],
        >>>     dates=my_dates,
        >>>     offsets=1,
        >>>     roll="forward",
        >>> ).tolist()
        >>> df.index = pd.to_datetime(session_end)

    """
    df = daily_df.copy()
    df.index = pd.to_datetime(df.index)
    df["stop"] = df.index + pd.to_timedelta(17, "h")
    df["start"] = df.stop - pd.to_timedelta(23, "h")
    df.set_index("start", inplace=True)
    df.drop(["stop"], axis=1, inplace=True)
    return df


def trading_holidays_in_range(start, end):
    """Returns a list of dates where CME holidays fall on a weekday

    Allow certain days, like 2018 National Day of Mourning to trade

    Holiday schedule: https://www.cmegroup.com/tools-information/holiday-calendar.html

    Args:
        start (str): YYYY-MM-DD
        end (str): YYYY-MM-DD

    Returns:
        list: list of dates in YYYY-MM-DD corresponding to holidays

    """

    # Add a week of buffer to catch any holidays upcoming
    end_with_buffer = (pd.to_datetime(end) + pd.Timedelta("7 days")).strftime(
        "%Y-%m-%d"
    )

    ALLOWED_HOLIDAYS_FOR_TRADING = [
        "2018-12-05",  # National Day of Mourning
    ]

    INVALID_HOLIDAYS_FOR_TRADING = []

    # get valid trading days per the module
    all_weekdays = pd.date_range(start, end_with_buffer, freq="B")
    cme = mcal.get_calendar("CME")
    schedule = cme.schedule(start_date=all_weekdays[0], end_date=all_weekdays[-1])
    valid_trading_days = pd.to_datetime(schedule["market_close"].dt.date)

    # remove invalid days (corrections to the module)
    for day in INVALID_HOLIDAYS_FOR_TRADING:
        try:
            valid_trading_days.drop(pd.to_datetime(day), inplace=True)
        except KeyError:
            pass

    # Create holidays
    holidays = np.setdiff1d(all_weekdays, valid_trading_days)
    holidays = [pd.to_datetime(h).strftime("%Y-%m-%d") for h in holidays]

    # add early closes to holidays, which is true except for Christmas Eve and Thanksgiving Eve
    early_closes = [
        x.strftime("%Y-%m-%d")
        for x in cme.early_closes(schedule=schedule).index.to_list()
    ]
    # Allow some trading on the day before or after Thanksgiving (i.e. if
    # it is in November and not Thursday)
    for date in early_closes:
        year, month, day = date.split("-")
        if month == "11":
            if pd.to_datetime(date).day_name() != "Thursday":
                early_closes.remove(date)
    for date in early_closes:
        year, month, day = date.split("-")
        if month == "12" and day == "24":
            early_closes.remove(date)

    # combine them
    holidays.extend(early_closes)
    holidays.sort()

    # Allow some trading on Christmas Eve
    dates_to_remove = []
    for holiday in holidays:
        year, month, day = holiday.split("-")
        if month == "12" and day == "24":
            logger.debug(f"found {holiday}")
            dates_to_remove.append(date)
    for date in dates_to_remove:
        try:
            holidays.remove(date)
        except ValueError:
            pass

    # add missing holidays (corrections to the module)
    for day in ALLOWED_HOLIDAYS_FOR_TRADING:
        try:
            holidays.remove(day)
        except ValueError:
            pass

    return holidays


def tradingday_offset(start, end, **kwargs):
    """Uses a trading day offset with CME calendar instead of business day offset

    Args:
        start (str): YYYY-MM-DD
        end (str): YYYY-MM-DD
        **kwargs: passed to np.busday_offset, requires "dates" "offsets" and optional is "roll"

    Returns:
        list of pd.Timestamps: dates YYYY-MM-DD timestamps
    """
    holidays = trading_holidays_in_range(start, end)
    offsets = np.busday_offset(
        **kwargs, holidays=holidays, weekmask="Mon Tue Wed Thu Fri"
    )
    offsets = pd.to_datetime(offsets)
    return offsets
